### PR TITLE
Adding support for inherent reactions

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
When we original designed the IDAES framework, we considered reactions to be isolated to specific reactor units, thus the use of specialized Reaction Blocks to hold reaction constraints. However, this does not work so well for electrolyte systems where many reactions are inherent in the system and need to be considered in all unit operations.

However, many unit models do not support reactions (i.e. most do not), so we need a new approach for dealing with this.

This PR proposes a solution to this - adding a new type of reaction ("inherent reactions") that need to be present in all unit operation and that are contained within the thermophysical property package instead of the reaction package (similar to phase equilibria). This adds a new optional term in the control volume material balances, which is activated automatically by a flag in the property package if inherent reactions are present.

## Changes proposed in this PR:
- Add `has_inherent_reactions` property to `PhysicalParameterBlocks`, and methods to access this from the `StateBlocks`.
- Add new terms to control volume material balances and scaling factors
- Tests for new attributes and terms
- No documentation as yet....

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by andrewlee94** [comment]
I am open to suggests on alternative approaches for this. I am a little concerned about adding a new distinction in types of reactions, but I haven't thought of a better approach yet.

**Review by eslickj — APPROVED** [review]
It all looks good to me, but should probably get one review from someone with immediate plans to use this if possible to make sure everything makes sense.

**Comment by andrewlee94** [comment]
@eslickj Unfortunately the only person immediately planning to use this is me....

NAWI will need this eventually, but I am not sure when another person will get to this.

**Comment by eslickj** [comment]
@andrewlee94, in that case, I'll take a closer look.

**Comment by eslickj** [comment]
It still seems good to me.  The only other idea I have (which may not be a good one) is that if you used a sort of true and apparent component approach, you could hide this from the framework and handle the inherent reaction business entirely in the property package, since framework would see apparent components, so the reactions wouldn't affect the mass balances there, and the reaction balances could go with the property package.  I guess the down side is things like separation units would need to be specially design to see the true components, so maybe that approach doesn't really get you anything better.

**Comment by andrewlee94** [comment]
@eslickj I am working to support both true and apparent species for the material baalnces, but the issue comes when you also have kinetic reactions. Here you need to have a term in the material balances, but you would need to write these in terms of apparent species too which is not necessarily easy.

**Comment by andrewlee94** [comment]
@eslickj's comments have made me realize that this might need more work, given different behaviour between true and apparent species.

**Comment by adowling2** [comment]
@andrewlee94 Would this framework allow for certain reactions to be added to all unit operations (control volumes)? Would it be possible to use this approach to make the reactor models (PBR, PRF, CSTR, etc.) a special case of this generalized approach? This is more of a long-term big picture question.

**Comment by andrewlee94** [comment]
@adowling2 This uses the same approach that is currently used for other types of reactions, then only difference being the logic used to determine if a given type of reaction should be included.

**Comment by andrewlee94** [comment]
OK, this got a lot bigger quickly, but it now works and has tests to prove it.

**Comment by andrewlee94** [comment]
Additional note: I found the source of the unexpected degree of freedom when testing inherent reactions in 1D systems, however that needs a separate PR (#230), as it has some (hopefully minor) wider reaching effects.

File: idaes/core/tests/test_control_volume_0d.py [hunk 1]
```diff
@@ -729,6 +731,53 @@ def test_add_phase_component_balances_eq_rxns_no_idx():
...
 
 
+@pytest.mark.unit
+def test_add_phase_component_balances_in_rxns():
+    m = ConcreteModel()
+    m.fs = Flowsheet(default={"dynamic": False})
+    m.fs.pp = PhysicalParameterTestBlock()
+
+    # Set property package to contain inherent reactions
+    m.fs.pp._has_inherent_reactions = True
```
Comment (TimBartholomew) [inline idaes/core/tests/test_control_volume_0d.py:741]: In practice, this `_has_inherent_reactions` attribute would be set within the property package, correct?

In other words, the user wouldn't necessarily set this at the flowsheet level, right?

**Review by TimBartholomew — APPROVED** [review]
LGTM. The approach looks good, I'm assuming the user testing and needs for additional features will come up with your interactions with the NAWI chemistry subteam.

File: idaes/core/tests/test_control_volume_0d.py (hunk 1 above)
Comment (TimBartholomew) [reply idaes/core/tests/test_control_volume_0d.py:741]: Oh I thought I got rid of this comment. I saw you set the attribute in the build method of the generic property package...

Comment (andrewlee94) [reply idaes/core/tests/test_control_volume_0d.py:741]: To answer anyway: Yes, this set by the property package. This was just for generic testing purposes.

**Review by eslickj — APPROVED** [review]
Looks good to me too.  This is one of those things I don't think I can get a good grip on without really using it, but I see no problems.

**Pull request merged**